Sriwan Khaomuang

Project 3

This project is to analyzed data about the App Store and Google Play mobile apps with the goal of recommending an app profile that can be profitable for both markets.

In [1]:
from csv import reader

opened_file = open('GooglePlayStore2018.csv')
read_file = reader(opened_file)
android = list(read_file)
android_header = android[0]
android = android[1:]

In [2]:
opened_file = open('AppleStore2017.csv')
read_file = reader(opened_file)
ios = list(read_file)
ios_header = ios[0]
ios = ios[1:]

In [3]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]
    for row in dataset_slice:
        print(row)
        print('\n')
    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

EXPLORE ANDROID DATASET

In [4]:
print(android_header)
print('\n')
explore_data(android, 0, 3, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', '7-Jan-18', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', '15-Jan-18', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', '1-Aug-18', '1.2.4', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


EXPLORE IOS DATASET

In [5]:
print(ios_header)
print('\n')
explore_data(ios, 0, 3, True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0', '2974676', '212', '3.5', '3.5', '95', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0', '2161558', '1289', '4.5', '4', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 7197
Number of columns: 16


CHECK IF ALL THE APP HAVE RATING BETWEEN 0-5

In [6]:
for app in android:
    rating = app[2]
    if float(rating) < 0 or float(rating) > 5:
        print(app)

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', '11-Feb-18', '1.0.19', '4.0 and up', '']


From the discussion section there's a wrong rating in row 10472, so delete that app

In [7]:
del(android[10472])
print('Total android app:', len(android))

Total android app: 10840


GET UNIQUE AND DUPLICATE ANDROID APPS

In [8]:
duplicate_apps = []
unique_apps = []

for app in android:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)

print('Number of duplicate apps:', len(duplicate_apps))
print('\n')
print('Examples of duplicate apps:', duplicate_apps[:5])

Number of duplicate apps: 1181


Examples of duplicate apps: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings']


In [40]:
print('Expected length android:', len(android) - 1181)

Expected length android: 9659


GET UNIQUE AND DUPLICATE IOS APPS

In [41]:
duplicate_ios_apps = []
unique_ios_apps = []

for app in ios:
    name = app[1]
    if name in unique_ios_apps:
        duplicate_ios_apps.append(name)
    else:
        unique_ios_apps.append(name)

print('Number of duplicate ios apps:', len(duplicate_ios_apps))
print('\n')
print('Examples of duplicate ios apps:', duplicate_ios_apps[:5])

Number of duplicate ios apps: 2


Examples of duplicate ios apps: ['Mannequin Challenge', 'VR Roller Coaster']


In [39]:
print('Expected length ios:', len(ios) - 2)

Expected length: 7195


GET ONLY THE HIGHEST REVIEW ANDROID APP

In [42]:
reviews_max = {}
for app in android:
    name = app[0]
    n_reviews = float(app[3])

    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews

    elif name not in reviews_max:
        reviews_max[name] = n_reviews

In [43]:
print('Expected length:', len(android) - 1181)
print('Actual length:', len(reviews_max))

Expected length: 9659
Actual length: 9659


USE DICTIONARY TO REMOVE DUPLICATE APP

In [21]:
cleaned_android = []
already_added = []

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    if (reviews_max[name] == n_reviews) and (name not in already_added):
        cleaned_android.append(app)
        already_added.append(name)

print('Example and number of cleaned data:')
print('\n')
explore_data(cleaned_android, 0, 3, True)


Example and number of cleaned data:


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', '7-Jan-18', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', '1-Aug-18', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', '8-Jun-18', 'Varies with device', '4.2 and up']


Number of rows: 9659
Number of columns: 13


GET ONLY THE HIGHEST REVIEW IOS APP

In [44]:
reviews_max_ios = {}
for app in ios:
    name = app[1]
    n_reviews = float(app[5])

    if name in reviews_max_ios and reviews_max_ios[name] < n_reviews:
        reviews_max_ios[name] = n_reviews

    elif name not in reviews_max_ios:
        reviews_max_ios[name] = n_reviews

In [45]:
print('Expected length:', len(ios) - 2)
print('Actual length:', len(reviews_max_ios))

Expected length: 7195
Actual length: 7195


In [47]:
cleaned_ios = []
already_added = []

for app in ios:
    name = app[1]
    n_reviews = float(app[5])
    if (reviews_max_ios[name] == n_reviews) and (name not in already_added):
        cleaned_ios.append(app)
        already_added.append(name)

print('Example and number of cleaned ios data:')
print('\n')
explore_data(cleaned_ios, 0, 3, True)


Example and number of cleaned ios data:


['284882215', 'Facebook', '389879808', 'USD', '0', '2974676', '212', '3.5', '3.5', '95', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0', '2161558', '1289', '4.5', '4', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 7195
Number of columns: 16


CREATE A FUNCTION TO GET ENGLISH APP
When there's more than 3 non-english letters, we'll consider as non-english app

In [18]:
def english(string):
    non_ascii = 0
    for each in string:
        if ord(each) > 127:
            non_ascii += 1

    if non_ascii > 3:
        return False
    else:
        return True

TEST FUNCTION

In [19]:
print(english('instagram'))
print(english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(english('Docs To Go™ Free Office Suite'))
print(english('Instachat 😜'))

True
False
True
True


GET ENGLISH APP OR ANDROID AND IOS

In [48]:
googleplay_english = []
appstore_english = []

for app in cleaned_android:
    name = app[0]
    if english(name):
        googleplay_english.append(app)

for app in cleaned_ios:
    name = app[1]
    if english(name):
        appstore_english.append(app)

explore_data(googleplay_english, 0, 3, True)
print('\n')
explore_data(appstore_english, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', '7-Jan-18', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', '1-Aug-18', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', '8-Jun-18', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13


['284882215', 'Facebook', '389879808', 'USD', '0', '2974676', '212', '3.5', '3.5', '95', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0', '2161558', '1289', '4.5', '4', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '

REMOVE APPS THAT AREN'T FREE
Only the one that price = 0

In [49]:
appstore_free = []
googleplay_free = []

for app in appstore_english:
    app_price = app[4]
    if app_price == '0':
        appstore_free.append(app)

for app in googleplay_english:
    app_price = app[7]
    if app_price == '0':
        googleplay_free.append(app)


print('android free app:',len(googleplay_free))
print('ios free app:',len(appstore_free))

android free app: 8864
ios free app: 3220


Create a function for generating frequency table and display

In [50]:
def freq_table(dataset, index):
    table = {}
    total = 0

    for row in dataset:
        total += 1
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1

    table_percentages = {}
    for key in table:
        percentage = (table[key] / total) * 100
        table_percentages[key] = percentage

    return table_percentages


def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

Analyze the frequency table for prime_genre of the App Store data set

In [31]:
display_table(appstore_free, -5)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


The most common genre for App Store is Games comprising more than 50% of all apps. Second place fell to Entertainment at 7.9%, follow by Photo&Video at 5% and Education with 3.66% respectively. The general impression is that the educational, productivity, and utility apps are not as commonly developed for the appstore as gaming, entertainment, and photo & video apps. Even though Games have a lot larger slice than other apps, that doesn't truly represent the large number of users.

Analyze the frequency table for prime_genre of the App Store data set

In [51]:
display_table(googleplay_free, 1)

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

The most common category for Play Store is Family with 19%. Second place fell to Game at 9.7%, follow by Tools at 8.5% and Business with 4.6% respectively. It also appears that apps for family, tools, business, lifestyle, productivity are more popular in the google play store and practical apps have a bigger share than in App Store market.

GET THE MOST POPULAR APP FOR IOS

In [55]:
genres_ios = freq_table(appstore_free, -5)

for genre in genres_ios:
    total = 0
    len_genre = 0
    for app in appstore_free:
        genre_app = app[-5]
        if genre_app == genre:
            n_ratings = float(app[5])
            total += n_ratings
            len_genre += 1
    avg_n_ratings = total / len_genre
    print(genre, ':', avg_n_ratings)

Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22812.92467948718
Music : 57326.530303030304
Reference : 74942.11111111111
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 18684.456790123455
Travel : 28243.8
Shopping : 26919.690476190477
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 39758.5
Finance : 31467.944444444445
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


According to the result navigation apps are the most popular genre with 86,090 average rating, follow by reference at around 75,000 and social networking at 72,000 total ratings

SEE NAME AND TOTAL NUMBER OF USER RATING FOR THE TOP 3 APPS

In [59]:
for app in appstore_free:
    name = app[1]
    genre = app[-5]
    rating_tot = app[5]
    if genre == 'Navigation':
        print(name, rating_tot)

Waze - GPS Navigation, Maps & Real-time Traffic 345046
Google Maps - Navigation & Transit 154911
Geocaching® 12811
CoPilot GPS – Car Navigation & Offline Maps 3582
ImmobilienScout24: Real Estate Search in Germany 187
Railway Route Search 5


In [60]:
for app in appstore_free:
    name = app[1]
    genre = app[-5]
    rating_tot = app[5]
    if genre == 'Reference':
        print(name, rating_tot)

Bible 985920
Dictionary.com Dictionary & Thesaurus 200047
Dictionary.com Dictionary & Thesaurus for iPad 54175
Google Translate 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition 17588
Merriam-Webster Dictionary 16849
Night Sky 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools 4693
GUNS MODS for Minecraft PC Edition - Mods Tools 1497
Guides for Pokémon GO - Pokemon GO News and Cheats 826
WWDC 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free 718
VPN Express 14
Real Bike Traffic Rider Virtual Reality Glasses 8
教えて!goo 0
Jishokun-Japanese English Dictionary & Translator 0


In [63]:
for app in appstore_free:
    name = app[1]
    genre = app[-5]
    rating_tot = app[5]
    if genre == 'Social Networking':
        print(name, rating_tot)

Facebook 2974676
Pinterest 1061624
Skype for iPhone 373519
Messenger 351466
Tumblr 334293
WhatsApp Messenger 287589
Kik 260965
ooVoo – Free Video Call, Text and Voice 177501
TextNow - Unlimited Text + Calls 164963
Viber Messenger – Text & Call 164249
Followers - Social Analytics For Instagram 112778
MeetMe - Chat and Meet New People 97072
We Heart It - Fashion, wallpapers, quotes, tattoos 90414
InsTrack for Instagram - Analytics Plus More 85535
Tango - Free Video Call, Voice and Chat 75412
LinkedIn 71856
Match™ - #1 Dating App. 60659
Skype for iPad 60163
POF - Best Dating App for Conversations 52642
Timehop 49510
Find My Family, Friends & iPhone - Life360 Locator 43877
Whisper - Share, Express, Meet 39819
Hangouts 36404
LINE PLAY - Your Avatar World 34677
WeChat 34584
Badoo - Meet New People, Chat, Socialize. 34428
Followers + for Instagram - Follower Analytics 28633
GroupMe 28260
Marco Polo Video Walkie Talkie 27662
Miitomo 23965
SimSimi 23530
Grindr - Gay and same sex guys chat, meet

In terms of the app profile recommendation on App Store, I would suggest a reference app like something to do with books, dictionary or maybe a translation app. Reference apps are in the second most popular category, and also have less competition in the market. From the perspective above, even though navigation apps and social networking apps are in the top 3 most popular kinds of app, both of them have a really big gap in overall number of users. That means users use just a few apps in these categories, and there's just a small chance for other apps to break into these categories and achieve large scale market penetration.

GET THE MOST POPULAR APP FOR ANDROID

In [52]:
display_table(googleplay_free, 5)

1,000,000+ : 15.726534296028879
100,000+ : 11.552346570397113
10,000,000+ : 10.548285198555957
10,000+ : 10.198555956678701
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343


GET NUMBER OF INSTALLED APP BY CATEGORY FOR PLAY STORE

In [53]:
categories_android = freq_table(googleplay_free, 1)

for category in categories_android:
    total = 0
    len_category = 0
    for app in googleplay_free:
        category_app = app[1]
        if category_app == category:
            n_installs = app[5]
            n_installs = n_installs.replace(',', '')
            n_installs = n_installs.replace('+', '')
            total += float(n_installs)
            len_category += 1
    avg_n_installs = total / len_category
    print(category, ':', avg_n_installs)

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15588015.603248259
FAMILY : 3695641.8198090694
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_

According to the results, communication apps have the greatest market share with 38,456,119 installations. The runner-up is social at around 23 million, followed by games at 12 million, entertainment at 12 million, and book & reference at 8.8 million

In [58]:
for app in googleplay_free:
    name = app[0]
    category = app[1]
    install = app[5]
    if category == 'COMMUNICATION':
        print(name, install)

WhatsApp Messenger 1,000,000,000+
Messenger for SMS 10,000,000+
My Tele2 5,000,000+
imo beta free calls and text 100,000,000+
Contacts 50,000,000+
Call Free – Free Call 5,000,000+
Web Browser & Explorer 5,000,000+
Browser 4G 10,000,000+
MegaFon Dashboard 10,000,000+
ZenUI Dialer & Contacts 10,000,000+
Cricket Visual Voicemail 10,000,000+
TracFone My Account 1,000,000+
Xperia Link™ 10,000,000+
TouchPal Keyboard - Fun Emoji & Android Keyboard 10,000,000+
Skype Lite - Free Video Call & Chat 5,000,000+
My magenta 1,000,000+
Android Messages 100,000,000+
Google Duo - High Quality Video Calls 500,000,000+
Seznam.cz 1,000,000+
Antillean Gold Telegram (original version) 100,000+
AT&T Visual Voicemail 10,000,000+
GMX Mail 10,000,000+
Omlet Chat 10,000,000+
My Vodacom SA 5,000,000+
Microsoft Edge 5,000,000+
Messenger – Text and Video Chat for Free 1,000,000,000+
imo free video calls and chat 500,000,000+
Calls & Text by Mo+ 5,000,000+
free video calls and chat 50,000,000+
Skype - free IM & video

For the app profile recommendation on Play Store, I would recommend either an entertainment app or a book & reference app for the reason that both of them are in the top 5 most popular apps, but there is still comparatively less competition than in other categories. From the result above, even though communication app has the biggest chunk of the pie (as well as social which in the similar category), there is a big gap in number of users. There are just a few communication apps to have reached one billion users, while a lot of them there are just a few thousand downloads.